src/cli.ts

```diff
@@ -0,0 +1,145 @@
+  const s3Host = env.getS3HostOverride();
+  if (s3Host) {
+    // Override config value if set via envar
+    set(config, 'pjson.oclif.update.s3.host', s3Host);
+  }
+
+  const npmRegistry = env.getNpmRegistryOverride();
+  if (npmRegistry) {
+    // Override config value if set via envar
+    set(config, 'pjson.oclif.warn-if-update-available.registry', npmRegistry);
+  }
+}
+
+export function configureAutoUpdate(envars: Env): void {
+  if (envars.isDemoMode()) {
+    // Disable autoupdates in demo mode
+    envars.setAutoupdateDisabled(true);
+    envars.setUpdateInstructions(UPDATE_DISABLED_DEMO);
+    return;
+  }
+
+  if (envars.isInstaller()) {
+    envars.normalizeAutoupdateDisabled();
+    if (envars.isAutoupdateDisabled()) {
+      envars.setUpdateInstructions(UPDATE_DISABLED_INSTALLER);
+    }
+    return;
+  }
+
+  // Not an installer, so this must be running from an npm installation
+  if (!envars.isAutoupdateDisabledSet()) {
+    // Disable autoupdates if run from an npm install or in local dev, if not explicitly set
+    envars.setAutoupdateDisabled(true);
+  }
+
+  if (envars.isAutoupdateDisabled()) {
+    envars.setUpdateInstructions(UPDATE_DISABLED_NPM);
+  }
+}
+
+function debugCliInfo(version: string, channel: string, env: Env, config: Interfaces.Config): void {
+  function debugSection(section: string, items: Array<[string, string]>): void {
+    const pad = 25;
+    debug('%s:', section.padStart(pad));
+    items.forEach(([name, value]) => debug('%s: %s', name.padStart(pad), value));
+  }
+
+  debugSection('OS', [
+    ['platform', os.platform()],
+    ['architecture', os.arch()],
+    ['release', os.release()],
+    ['shell', config.shell],
+  ]);
+
+  debugSection('NODE', [['version', process.versions.node]]);
+
+  debugSection('CLI', [
+    ['version', version],
+    ['channel', channel],
+    ['bin', config.bin],
+    ['data', config.dataDir],
+    ['cache', config.cacheDir],
+    ['config', config.configDir],
+  ]);
+
+  debugSection(
+    'ENV',
+    [...envVars].map((key): [string, string] => [key, env.getString(key, '<not set>')] as [string, string])
+  );
+
+  debugSection(
+    'ARGS',
+    process.argv.map((arg, i): [string, string] => [i.toString(), arg])
+  );
+}
+
+export function create(
+  version: string,
+  channel: string,
+  run: typeof oclifRun,
+  env = nodeEnv
+): { run: () => Promise<unknown> } {
+  const root = path.resolve(__dirname, '..');
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const pjson = require(path.resolve(__dirname, '..', 'package.json')) as AnyJson;
+  const args = process.argv.slice(2);
+
+  return {
+    async run(): Promise<unknown> {
+      const config = new Config({ name: get(pjson, 'oclif.bin') as string | undefined, root, version, channel });
+      await config.load();
+      configureUpdateSites(config, env);
+      configureAutoUpdate(env);
+      debugCliInfo(version, channel, env, config);
+      // I think the run method is used in test.
```

Comment:
Can we change this comment to link to where/how `run` is used in tests?
[example](https://github.com/salesforcecli/cli/pull/171/files#diff-1deee0a575599b2df117c280da319f7938aaf6fdb0c04bcdbde769dbf464be69R46)